lib.packagesFromDirectoryRecursive: refactor (v2)

nicoonoclaste opened the pull request:
Second go at #359941

An assertion was too restrictive and broke the installer's tests, due to `directory` being a symlink.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- Tested, as applicable:
  - [x] [NixOS test](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) `installer`
  - [x] `nix-instantiate --eval --strict lib/tests/misc.nix`
  - [x] `nix-build lib/tests/release.nix`
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nicoonoclaste committed:
lib.packagesFromDirectoryRecursive: refactor, again

This is a second take on #359941, which was reverted by
  940db5766a966b5586b53359bb785badac6c5662.

Co-authored-by: Silvan Mosberger <github@infinisil.com>

K900 commented:
Ofborg's lib tests can help here.

K900 commented:
You want to run `lib/tests/release.nix`.

K900 commented:
"Blocker" generally refers to things that are blocking major changes, not feature additions.

infinisil commented on lib/filesystem.nix:
```diff
@@ -361,51 +364,28 @@ in
+      defaultPath = append directory "package.nix";
     in
-    packagesFromDirectory directory;
+    if pathIsRegularFile defaultPath then
```
There is actually another change here: This was just `pathExists` before, while now it's `pathIsRegularFile`. Because `pathIsRegularFile` returns `false` on symlinks, this could break some use cases, so we should keep this the same as before

```suggestion
    if pathExists defaultPath then
```

infinisil commented on lib/filesystem.nix:
```diff
@@ -361,51 +364,28 @@ in
+    if pathIsRegularFile defaultPath then
+      # if `${directory}/package.nix` exists, call it directly
+      callPackage defaultPath {}
+    else lib.concatMapAttrs (name: type:
```
```suggestion
    else concatMapAttrs (name: type:
```

And add `concatMapAttrs` to the import list at the top. A bit annoying, but this is the standard to keep `lib` more performant.

infinisil reviewed (changes requested):
Didn't have the chance to review this the first time, so here's a late one. Just minor things though, looking good overall :)

nicoonoclaste committed:
lib.packagesFromDirectoryRecursive: let-bind functions called from `lib`

I was told it is better for eval performance:
  https://github.com/NixOS/nixpkgs/pull/361424#discussion_r1879975642

infinisil commented on lib/filesystem.nix:
```diff
@@ -361,51 +364,29 @@ in
-        else mapAttrs'
-          (directoryEntryToAttrPair path)
-          (packageDirectoryEntries path);
+      inherit (lib) concatMapAttrs removeSuffix;
+      inherit (lib.path) append;
```
Usually this should be added to the very top of this file, but after thinking about it, the performance benefit is also there if you do it here. In any case, that's a tiny detail and not worth nitpicking over further :)

infinisil reviewed (approved):
LGTM! Did you want to squash yourself? Feel free to merge yourself as you wish :)

phanirithvij commented on lib/filesystem.nix:
```diff
@@ -361,51 +364,29 @@ in
-          (packageDirectoryEntries path);
+      inherit (lib) concatMapAttrs removeSuffix;
+      inherit (lib.path) append;
+      defaultPath = append directory "package.nix";
```
Previously `directory` could be a string but this forces it to be a path.
But I guess this change is heading in the right direction and users should adapt?

for e.g. https://github.com/9999years/git-prole/blob/68b4ba943a4bbb3d8452c67bbf50b0ba3f58956a/nix/makePackages.nix#L12

9999years replied:
This seems like a pretty pointless limitation that makes it very hard to interface with various parts of Nix, which is pretty lackadaisical about what it considers to be a path:

- `outPath` on derivations and Flake inputs is not a path.
- Derivations can be coerced to their `outPath` in string interpolation, but that produces strings, not paths.
- `builtins.path`, bizarrely, returns a string instead of a path (not that the documentation makes this clear).

This is a frustrating change, especially since `lib.path.append` is defined as generic concatenation with `+` anyways, just with error messages where there weren't any before.

One reason I wanted to support strings with the original implementation is strings make it a lot easier to avoid footguns like https://github.com/NixOS/nix/issues/9428 which can (for industrial use cases) lead to gigabytes of data being uselessly copied around the Nix store. The fact that strings can be used with `builtins.readDir` and `builtins.pathExists` without copying paths to the Nix store is a feature, not a bug!
